cudax/include/cuda/experimental/__stf/internal/cuda_kernel_scope.cuh

```diff
@@ -45,46 +53,200 @@ class stream_ctx;
  */
 struct cuda_kernel_desc
 {
+  cuda_kernel_desc() = default;
+
   template <typename Fun, typename... Args>
   cuda_kernel_desc(Fun func, dim3 gridDim_, dim3 blockDim_, size_t sharedMem_, Args... args)
-      : func((const void*) func)
-      , gridDim(gridDim_)
-      , blockDim(blockDim_)
-      , sharedMem(sharedMem_)
+  {
+    configure(mv(func), gridDim_, blockDim_, sharedMem_, ::std::forward<Args>(args)...);
+  }
+
+  template <typename Fun, typename... Args>
+  void configure(Fun func, dim3 gridDim_, dim3 blockDim_, size_t sharedMem_, Args... args)
   {
     using TupleType = ::std::tuple<::std::decay_t<Args>...>;
 
+    _CCCL_ASSERT(!configured, "cuda_kernel_desc was already configured");
+
+    func_variant = store_func(mv(func));
+    gridDim      = gridDim_;
+    blockDim     = blockDim_;
+    sharedMem    = sharedMem_;
+
     // We first copy all arguments into a tuple because the kernel
     // implementation needs pointers to the argument, so we cannot use
     // directly those passed in the pack of arguments
-    auto arg_tuple = ::std::make_shared<TupleType>(std::forward<Args>(args)...);
+    auto arg_tuple = ::std::make_shared<TupleType>(mv(args)...);
 
-    // Ensure we are packing arguments of the proper types to call func
-    static_assert(::std::is_invocable_v<Fun, Args...>);
+    // Ensure we are packing arguments of the proper types to call func (only
+    // valid with the runtime API)
+    static_assert(reserved::is_function_or_kernel_v<Fun> || ::std::is_invocable_v<Fun, Args...>);
```

Comment:
guess this stuff could be moved to the top of the function, it's a precondition of sorts (doesn't make any difference)